File: src/cli/app/asset/audit.py
#!/usr/bin/env python3
"""
Git 仓库资产审计模块

根据 docs/handbook/asset/governace/git_repo.md 规范，
检查 Git 仓库是否符合标准资产体系要求。
"""

import re
import subprocess
import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

import typer


@dataclass
class AuditResult:
    """审计结果"""

    name: str
    passed: bool
    message: str
    suggestion: Optional[str] = None


@dataclass
class AuditReport:
    """审计报告"""

    repo_path: str
    results: list[AuditResult] = field(default_factory=list)

    @property
    def passed_count(self) -> int:
        return sum(1 for r in self.results if r.passed)

    @property
    def failed_count(self) -> int:
        return sum(1 for r in self.results if not r.passed)

    @property
    def total_count(self) -> int:
        return len(self.results)

    @property
    def pass_rate(self) -> float:
        if self.total_count == 0:
            return 0.0
        return self.passed_count / self.total_count * 100

    def print_report(self, verbose: bool = False):
        """打印审计报告"""
        print("\n" + "=" * 60)
        print("Git 仓库资产审计报告")
        print("=" * 60)
        print(f"仓库路径：{self.repo_path}")
        print(f"审计结果：{self.passed_count}/{self.total_count} 通过 "
              f"({self.pass_rate:.1f}%)")
        print("-" * 60)

        # 先显示未通过的项目
        failed_results = [r for r in self.results if not r.passed]
        if failed_results:
            print("\n❌ 未通过项目:")
            for result in failed_results:
                print(f"\n  [{result.name}]")
                print(f"  {result.message}")
                if result.suggestion:
                    print(f"  💡 建议：{result.suggestion}")

        # 显示通过的项目
        if verbose:
            passed_results = [r for r in self.results if r.passed]
            if passed_results:
                print("\n✅ 通过项目:")
                for result in passed_results:
                    print(f"  ✓ {result.name}")

        print("\n" + "=" * 60)

        if self.failed_count > 0:
            print("⚠️  审计未通过，请根据建议修复问题")
            return False
        else:
            print("✅ 审计通过，仓库符合标准资产体系规范")
            return True


class GitRepoAuditor:
    """Git 仓库审计器"""

    REQUIRED_FILES = {
        "README.md": "项目概述、目录结构",
        "CONTRIBUTING.md": "贡献指南、工作流、环境变量",
        "AGENTS.md": "Agent 导航",
        "CHANGELOG.md": "版本历史",
        ".gitignore": "Git 忽略规则",
    }

    OPTIONAL_DIRS = {
        "meta": "元数据目录",
    }

    COMMIT_TYPES = {
        "feat", "fix", "docs", "test",
        "refactor", "chore", "style", "perf"
    }

    def __init__(self, repo_path: str):
        self.repo_path = Path(repo_path).resolve()
        self._results: list[AuditResult] = []

    def audit(self) -> AuditReport:
        """执行完整审计"""
        if not self.repo_path.exists():
            print(f"错误：路径不存在 - {self.repo_path}")
            sys.exit(1)

        if not (self.repo_path / ".git").exists():
            print(f"错误：不是 Git 仓库 - {self.repo_path}")
            sys.exit(1)

        # 执行各项检查
        self._check_required_files()
        self._check_optional_dirs()
        self._check_readme_content()
        self._check_contributing_content()
        self._check_agents_content()
        self._check_changelog_format()
        self._check_gitignore_content()
        self._check_submodules()
        self._check_recent_commits()

        report = AuditReport(str(self.repo_path))
        report.results = self._results
        return report

    def _add_result(self, result: AuditResult):
        """添加审计结果"""
        self._results.append(result)

    def _check_required_files(self):
        """检查必需文件是否存在"""
        for filename, description in self.REQUIRED_FILES.items():
            file_path = self.repo_path / filename
            passed = file_path.exists()
            self._add_result(AuditResult(
                name=f"必需文件：{filename}",
                passed=passed,
                message=f"{filename} - {description}" if passed else f"缺少 {filename}",
                suggestion=f"创建 {filename} 文件" if not passed else None
            ))

    def _check_optional_dirs(self):
        """检查可选目录"""
        for dirname, description in self.OPTIONAL_DIRS.items():
            dir_path = self.repo_path / dirname
            passed = dir_path.exists() and dir_path.is_dir()
            self._add_result(AuditResult(
                name=f"可选目录：{dirname}/",
                passed=passed,
                message=f"{dirname}/ - {description}" if passed else f"缺少 {dirname}/ 目录",
                suggestion=f"考虑创建 {dirname}/ 目录用于存储元数据" if not passed else None
            ))

    def _check_readme_content(self):
        """检查 README.md 内容"""
        readme_path = self.repo_path / "README.md"
        if not readme_path.exists():
            return

        content = readme_path.read_text(encoding="utf-8")

        # 检查是否包含项目简介
        has_intro = len(content.split("\n")[0].replace("#", "").strip()) > 0

        # 检查是否包含目录结构
        has_structure = "目录" in content or "结构" in content or "```" in content

        # 检查是否包含快速开始
        has_quickstart = ("快速" in content or "开始" in content or
                         "Quick" in content or "Start" in content or
                         "开始使用" in content)

        passed = has_intro and (has_structure or has_quickstart)
        self._add_result(AuditResult(
            name="README.md 内容规范",
            passed=passed,
            message="包含项目简介、目录结构、快速开始" if passed else "内容不完整",
            suggestion="添加项目简介、目录结构和快速开始指南" if not passed else None
        ))

    def _check_contributing_content(self):
        """检查 CONTRIBUTING.md 内容"""
        contrib_path = self.repo_path / "CONTRIBUTING.md"
        if not contrib_path.exists():
            return

        content = contrib_path.read_text(encoding="utf-8")

        # 检查关键章节
        required_sections = [
            ("项目结构", ["结构", "目录", "Project Structure"]),
            ("开发环境", ["开发", "环境", "Environment", "Setup"]),
            ("提交规范", ["提交", "Commit", "规范"]),
            ("发布流程", ["发布", "Release", "版本"]),
        ]

        missing_sections = []
        for section_name, keywords in required_sections:
            has_section = any(kw in content for kw in keywords)
            if not has_section:
                missing_sections.append(section_name)

        passed = len(missing_sections) == 0
        self._add_result(AuditResult(
            name="CONTRIBUTING.md 内容规范",
            passed=passed,
            message="包含项目结构、开发环境、提交规范、发布流程" if passed
                    else f"缺少章节：{', '.join(missing_sections)}",
            suggestion=f"添加缺失的章节：{', '.join(missing_sections)}" if not passed else None
        ))

    def _check_agents_content(self):
        """检查 AGENTS.md 内容"""
        agents_path = self.repo_path / "AGENTS.md"
        if not agents_path.exists():
            return

        content = agents_path.read_text(encoding="utf-8")
        lines = content.strip().split("\n")

        # 检查行数（建议 ~50 行）
        line_count = len(lines)
        is_concise = line_count <= 100  # 宽松一点，不超过 100 行

        # 检查是否包含使用场景表格
        has_table = "|" in content and "---" in content

        # 检查是否包含快速索引
        has_index = ("索引" in content or "Index" in content or
                    "README" in content or "CONTRIBUTING" in content)

        passed = is_concise and (has_table or has_index)
        self._add_result(AuditResult(
            name="AGENTS.md 内容规范",
            passed=passed,
            message=f"简洁 ({line_count}行)，包含使用场景和快速索引" if passed
                    else f"需要优化 (共{line_count}行)",
            suggestion="保持简洁 (~50 行)，添加使用场景表格和快速索引" if not passed else None
        ))

    def _check_changelog_format(self):
        """检查 CHANGELOG.md 格式"""
        changelog_path = self.repo_path / "CHANGELOG.md"
        if not changelog_path.exists():
            return

        content = changelog_path.read_text(encoding="utf-8")

        # 检查基本格式
        has_changelog_header = "# Changelog" in content or "# CHANGELOG" in content

        # 检查是否有版本记录
        has_version = bool(re.search(r'## \[?v?\d+\.\d+\.\d+', content))

        # 检查是否有分类标题
        has_sections = any(section in content for section in
                          ["### Added", "### Changed", "### Fixed", "### Removed"])

        passed = has_changelog_header and has_version
        self._add_result(AuditResult(
            name="CHANGELOG.md 格式规范",
            passed=passed,
            message="符合语义化版本格式" if passed else "格式不规范",
            suggestion="添加 # Changelog 标题和版本号，使用 ### Added/Changed/Fixed/Removed 分类"
                      if not passed else None
        ))

    def _check_gitignore_content(self):
        """检查 .gitignore 内容"""
        gitignore_path = self.repo_path / ".gitignore"
        if not gitignore_path.exists():
            return

        content = gitignore_path.read_text(encoding="utf-8")

        # 检查是否包含常见忽略规则
        common_patterns = [
            (".venv", "Python 虚拟环境"),
            ("__pycache__", "Python 缓存"),
            ("*.pyc", "Python 编译文件"),
            (".env", "环境变量文件"),
        ]

        found_patterns = []
        for pattern, description in common_patterns:
            if pattern in content:
                found_patterns.append(f"{pattern} ({description})")

        passed = len(found_patterns) >= 2  # 至少包含 2 个常见规则
        self._add_result(AuditResult(
            name=".gitignore 内容规范",
            passed=passed,
            message=f"包含 {len(found_patterns)} 个常见规则" if passed else "规则较少",
            suggestion="添加常见的忽略规则：.venv, __pycache__, *.pyc, .env 等"
                      if not passed else None
        ))

    def _check_submodules(self):
        """检查子模块配置"""
        gitmodules_path = self.repo_path / ".gitmodules"

        if not gitmodules_path.exists():
            self._add_result(AuditResult(
                name="子模块配置",
                passed=True,
                message="无子模块配置",
                suggestion=None
            ))
            return

        # 检查 .gitmodules 文件格式
        content = gitmodules_path.read_text(encoding="utf-8")
        has_submodule = "[submodule" in content

        # 检查子模块是否已推送（如果有远程）
        try:
            result = subprocess.run(
                ["git", "submodule", "status"],
                cwd=self.repo_path,
                capture_output=True,
                text=True,
                timeout=10
            )
            submodule_status = result.stdout.strip()

            # 检查是否有未推送的提交
            unpushed = False
            if submodule_status:
                for line in submodule_status.split("\n"):
                    if line.startswith("-") or line.startswith("+"):
                        unpushed = True
                        break

            passed = has_submodule and not unpushed
            self._add_result(AuditResult(
                name="子模块状态",
                passed=passed,
                message="子模块配置正确且已推送" if passed else "子模块有未推送的提交",
                suggestion="请先推送所有子模块的提交，再推送父仓库" if not passed else None
            ))
        except (subprocess.TimeoutExpired, Exception) as e:
            self._add_result(AuditResult(
                name="子模块状态",
                passed=has_submodule,
                message=f"子模块配置存在，状态检查跳过 ({e})",
                suggestion=None
            ))

    def _check_recent_commits(self):
        """检查最近的提交是否符合规范"""
        try:
            result = subprocess.run(
                ["git", "log", "--oneline", "-10"],
                cwd=self.repo_path,
                capture_output=True,
                text=True,
                timeout=10
            )

            if result.returncode != 0:
                return

            commits = result.stdout.strip().split("\n")
            if not commits:
                return

            # 检查提交信息格式
            conventional_pattern = re.compile(
                r'^[a-z]+\([a-z-]+\)?:|^feat:|^fix:|^docs:|^test:|^refactor:|^chore:|^style:|^perf:'
            )

            compliant_count = 0
            for commit in commits:
                # 跳过空行
                if not commit.strip():
                    continue
                # 提取提交信息（去掉 hash）
                message = commit.split(" ", 1)[1] if " " in commit else commit
                if conventional_pattern.match(message.lower()):
                    compliant_count += 1

            compliance_rate = compliant_count / len(commits) * 100 if commits else 0
            passed = compliance_rate >= 50  # 至少 50% 符合规范

            self._add_result(AuditResult(
                name="提交规范符合度",
                passed=passed,
                message=f"{compliant_count}/{len(commits)} 符合 Conventional Commits "
                       f"({compliance_rate:.0f}%)",
                suggestion="使用 `cz commit` 创建规范提交，或手动遵循 <type>: <description> 格式"
                          if not passed else None
            ))
        except (subprocess.TimeoutExpired, Exception) as e:
            self._add_result(AuditResult(
                name="提交规范符合度",
                passed=True,
                message=f"提交检查跳过 ({e})",
                suggestion=None
            ))


def audit(
    repo_path: str = typer.Argument(".", help="要审计的 Git 仓库路径"),
    verbose: bool = typer.Option(False, "--verbose", "-v", help="显示所有通过的项目")
) -> bool:
    """
    审计 Git 仓库是否符合标准资产体系规范
    
    检查项目包括：必需文件、可选目录、README/CONTRIBUTING/AGENTS/CHANGELOG 内容规范、
    .gitignore 规则、子模块状态、提交规范符合度
    
    Returns:
        是否通过审计
    """
    auditor = GitRepoAuditor(repo_path)
    report = auditor.audit()
    passed = report.print_report(verbose)
    if not passed:
        raise typer.Exit(code=1)
    return True

File: src/cli/src/asset/audit.rs
use std::path::{Path, PathBuf};

use clap::Args;
use regex::Regex;

use crate::git_utils;

#[derive(Args)]
pub struct AuditArgs {
    /// 要审计的 Git 仓库路径
    #[arg(default_value = ".")]
    pub repo_path: String,

    /// 显示所有通过的项目
    #[arg(short, long)]
    pub verbose: bool,
}

#[derive(Debug)]
pub struct AuditResult {
    pub name: String,
    pub passed: bool,
    pub message: String,
    pub suggestion: Option<String>,
}

pub struct AuditReport {
    pub repo_path: String,
    pub results: Vec<AuditResult>,
}

impl AuditReport {
    pub fn passed_count(&self) -> usize {
        self.results.iter().filter(|r| r.passed).count()
    }

    pub fn failed_count(&self) -> usize {
        self.results.iter().filter(|r| !r.passed).count()
    }

    pub fn total_count(&self) -> usize {
        self.results.len()
    }

    pub fn pass_rate(&self) -> f64 {
        if self.results.is_empty() {
            return 0.0;
        }
        self.passed_count() as f64 / self.total_count() as f64 * 100.0
    }

    pub fn print_report(&self, verbose: bool) -> bool {
        println!();
        println!("{}", "=".repeat(60));
        println!("Git 仓库资产审计报告");
        println!("{}", "=".repeat(60));
        println!("仓库路径：{}", self.repo_path);
        println!(
            "审计结果：{}/{} 通过 ({:.1}%)",
            self.passed_count(),
            self.total_count(),
            self.pass_rate()
        );
        println!("{}", "-".repeat(60));

        let failed: Vec<_> = self.results.iter().filter(|r| !r.passed).collect();
        if !failed.is_empty() {
            println!("\n❌ 未通过项目:");
            for result in &failed {
                println!("\n  [{}]", result.name);
                println!("  {}", result.message);
                if let Some(suggestion) = &result.suggestion {
                    println!("  💡 建议：{suggestion}");
                }
            }
        }

        if verbose {
            let passed: Vec<_> = self.results.iter().filter(|r| r.passed).collect();
            if !passed.is_empty() {
                println!("\n✅ 通过项目:");
                for result in &passed {
                    println!("  ✓ {}", result.name);
                }
            }
        }

        println!("\n{}", "=".repeat(60));

        if self.failed_count() > 0 {
            println!("⚠️  审计未通过，请根据建议修复问题");
            false
        } else {
            println!("✅ 审计通过，仓库符合标准资产体系规范");
            true
        }
    }
}

pub struct GitRepoAuditor {
    pub repo_path: PathBuf,
    results: Vec<AuditResult>,
}

impl GitRepoAuditor {
    const REQUIRED_FILES: &'static [(&'static str, &'static str)] = &[
        ("README.md", "项目概述、目录结构"),
        ("CONTRIBUTING.md", "贡献指南、工作流、环境变量"),
        ("AGENTS.md", "Agent 导航"),
        ("CHANGELOG.md", "版本历史"),
        (".gitignore", "Git 忽略规则"),
    ];

    const OPTIONAL_DIRS: &'static [(&'static str, &'static str)] =
        &[("meta", "元数据目录")];

    pub fn new(repo_path: &str) -> Self {
        Self {
            repo_path: PathBuf::from(repo_path).canonicalize().unwrap_or_else(|_| {
                PathBuf::from(repo_path)
            }),
            results: Vec::new(),
        }
    }

    pub fn audit(&mut self) -> AuditReport {
        if !self.repo_path.exists() {
            eprintln!("错误：路径不存在 - {}", self.repo_path.display());
            std::process::exit(1);
        }

        if !self.repo_path.join(".git").exists() {
            eprintln!("错误：不是 Git 仓库 - {}", self.repo_path.display());
            std::process::exit(1);
        }

        self.check_required_files();
        self.check_optional_dirs();
        self.check_readme_content();
        self.check_contributing_content();
        self.check_agents_content();
        self.check_changelog_format();
        self.check_gitignore_content();
        self.check_submodules();
        self.check_recent_commits();
        self.check_release_consistency();

        AuditReport {
            repo_path: self.repo_path.display().to_string(),
            results: std::mem::take(&mut self.results),
        }
    }

    fn add_result(&mut self, result: AuditResult) {
        self.results.push(result);
    }

    fn read_file(&self, path: &Path) -> Option<String> {
        std::fs::read_to_string(path).ok()
    }

    fn check_required_files(&mut self) {
        for (filename, description) in Self::REQUIRED_FILES {
            let file_path = self.repo_path.join(filename);
            let passed = file_path.exists();
            self.add_result(AuditResult {
                name: format!("必需文件：{filename}"),
                passed,
                message: if passed {
                    format!("{filename} - {description}")
                } else {
                    format!("缺少 {filename}")
                },
                suggestion: if !passed {
                    Some(format!("创建 {filename} 文件"))
                } else {
                    None
                },
            });
        }
    }

    fn check_optional_dirs(&mut self) {
        for (dirname, description) in Self::OPTIONAL_DIRS {
            let dir_path = self.repo_path.join(dirname);
            let passed = dir_path.exists() && dir_path.is_dir();
            self.add_result(AuditResult {
                name: format!("可选目录：{dirname}/"),
                passed,
                message: if passed {
                    format!("{dirname}/ - {description}")
                } else {
                    format!("缺少 {dirname}/ 目录")
                },
                suggestion: if !passed {
                    Some(format!("考虑创建 {dirname}/ 目录用于存储元数据"))
                } else {
                    None
                },
            });
        }
    }

    fn check_readme_content(&mut self) {
        let readme_path = self.repo_path.join("README.md");
        let content = match self.read_file(&readme_path) {
            Some(c) => c,
            None => return,
        };

        let lines: Vec<&str> = content.lines().collect();
        let has_intro = lines.first().map(|l| l.replace('#', "").trim().len() > 0).unwrap_or(false);
        let has_structure = content.contains("目录") || content.contains("结构") || content.contains("```");
        let has_quickstart = content.contains("快速")
            || content.contains("开始")
            || content.contains("Quick")
            || content.contains("Start")
            || content.contains("开始使用");

        let passed = has_intro && (has_structure || has_quickstart);
        self.add_result(AuditResult {
            name: "README.md 内容规范".to_string(),
            passed,
            message: if passed {
                "包含项目简介、目录结构、快速开始".to_string()
            } else {
                "内容不完整".to_string()
            },
            suggestion: if !passed {
                Some("添加项目简介、目录结构和快速开始指南".to_string())
            } else {
                None
            },
        });
    }

    fn check_contributing_content(&mut self) {
        let contrib_path = self.repo_path.join("CONTRIBUTING.md");
        let content = match self.read_file(&contrib_path) {
            Some(c) => c,
            None => return,
        };

        let required = [
            ("项目结构", &["结构", "目录", "Project Structure"] as &[&str]),
            ("开发环境", &["开发", "环境", "Environment", "Setup"]),
            ("提交规范", &["提交", "Commit", "规范"]),
            ("发布流程", &["发布", "Release", "版本"]),
        ];

        let mut missing = Vec::new();
        for (section_name, keywords) in &required {
            let has = keywords.iter().any(|kw| content.contains(kw));
            if !has {
                missing.push(*section_name);
            }
        }

        let passed = missing.is_empty();
        self.add_result(AuditResult {
            name: "CONTRIBUTING.md 内容规范".to_string(),
            passed,
            message: if passed {
                "包含项目结构、开发环境、提交规范、发布流程".to_string()
            } else {
                format!("缺少章节：{}", missing.join(", "))
            },
            suggestion: if !passed {
                Some(format!("添加缺失的章节：{}", missing.join(", ")))
            } else {
                None
            },
        });
    }

    fn check_agents_content(&mut self) {
        let agents_path = self.repo_path.join("AGENTS.md");
        let content = match self.read_file(&agents_path) {
            Some(c) => c,
            None => return,
        };

        let line_count = content.lines().count();
        let is_concise = line_count <= 50;
        let has_table = content.contains('|') && content.contains("---");
        let has_index = content.contains("索引")
            || content.contains("Index")
            || content.contains("README")
            || content.contains("CONTRIBUTING");
        let has_self_update = (content.contains("更新") && content.contains("AGENTS"))
            || (content.contains("维护") && content.contains("AGENTS"))
            || content.to_lowercase().contains("self-update")
            || content.to_lowercase().contains("how to update");

        let passed = is_concise && (has_table || has_index) && has_self_update;
        self.add_result(AuditResult {
            name: "AGENTS.md 内容规范".to_string(),
            passed,
            message: if passed {
                format!("简洁 ({line_count}行)，包含使用场景、快速索引和自我更新说明")
            } else {
                format!("需要优化 (共{line_count}行)")
            },
            suggestion: if !passed {
                Some("保持简洁 (~50 行)，添加使用场景表格、快速索引，以及「如何更新 AGENTS.md」的说明".to_string())
            } else {
                None
            },
        });
    }

    fn check_changelog_format(&mut self) {
        let changelog_path = self.repo_path.join("CHANGELOG.md");
        let content = match self.read_file(&changelog_path) {
            Some(c) => c,
            None => return,
        };

        let has_header = content.contains("# Changelog") || content.contains("# CHANGELOG");
        let has_version = Regex::new(r"## \[?v?\d+\.\d+\.\d+")
            .unwrap()
            .is_match(&content);
        let passed = has_header && has_version;
        self.add_result(AuditResult {
            name: "CHANGELOG.md 格式规范".to_string(),
            passed,
            message: if passed {
                "符合语义化版本格式".to_string()
            } else {
                "格式不规范".to_string()
            },
            suggestion: if !passed {
                Some("添加 # Changelog 标题和版本号，使用 ### Added/Changed/Fixed/Removed 分类".to_string())
            } else {
                None
            },
        });
    }

    fn check_gitignore_content(&mut self) {
        let gitignore_path = self.repo_path.join(".gitignore");
        let content = match self.read_file(&gitignore_path) {
            Some(c) => c,
            None => return,
        };

        let common = [
            (".venv", "Python 虚拟环境"),
            ("__pycache__", "Python 缓存"),
            ("*.pyc", "Python 编译文件"),
            (".env", "环境变量文件"),
        ];

        let found: Vec<_> = common
            .iter()
            .filter(|(pattern, _)| content.contains(pattern))
            .collect();

        let passed = found.len() >= 2;
        self.add_result(AuditResult {
            name: ".gitignore 内容规范".to_string(),
            passed,
            message: if passed {
                format!("包含 {} 个常见规则", found.len())
            } else {
                "规则较少".to_string()
            },
            suggestion: if !passed {
                Some("添加常见的忽略规则：.venv, __pycache__, *.pyc, .env 等".to_string())
            } else {
                None
            },
        });
    }

    fn check_submodules(&mut self) {
        let gitmodules_path = self.repo_path.join(".gitmodules");

        if !gitmodules_path.exists() {
            self.add_result(AuditResult {
                name: "子模块配置".to_string(),
                passed: true,
                message: "无子模块配置".to_string(),
                suggestion: None,
            });
            return;
        }

        let content = match self.read_file(&gitmodules_path) {
            Some(c) => c,
            None => return,
        };
        let has_submodule = content.contains("[submodule");

        let submodule_status = git_utils::run_git_cmd(
            &["submodule", "status"],
            &self.repo_path,
        );

        let passed = match submodule_status {
            Ok(out) => {
                let unpushed = out.lines().any(|line| line.starts_with('-') || line.starts_with('+'));
                has_submodule && !unpushed
            }
            Err(_) => has_submodule,
        };

        self.add_result(AuditResult {
            name: "子模块状态".to_string(),
            passed,
            message: if passed {
                "子模块配置正确且已推送".to_string()
            } else {
                "子模块有未推送的提交".to_string()
            },
            suggestion: if !passed {
                Some("请先推送所有子模块的提交，再推送父仓库".to_string())
            } else {
                None
            },
        });
    }

    fn check_recent_commits(&mut self) {
        let log_result = git_utils::run_git_cmd(&["log", "--oneline", "-10"], &self.repo_path);

        let commits = match log_result {
            Ok(out) => {
                let lines: Vec<String> = out
                    .lines()
                    .filter(|l| !l.trim().is_empty())
                    .map(|s| s.to_string())
                    .collect();
                if lines.is_empty() {
                    return;
                }
                lines
            }
            Err(_) => return,
        };

        let conventional_re =
            Regex::new(r"^(feat|fix|docs|test|refactor|chore|style|perf)(\([a-z0-9-]+\))?:\s.+")
                .unwrap();

        let compliant_count = commits
            .iter()
            .filter(|c| {
                let msg = c.split_once(' ').map(|(_, m)| m).unwrap_or(c.as_str());
                conventional_re.is_match(&msg.to_lowercase())
            })
            .count();

        let compliance_rate = compliant_count as f64 / commits.len() as f64 * 100.0;
        let passed = compliance_rate >= 50.0;

        self.add_result(AuditResult {
            name: "提交规范符合度".to_string(),
            passed,
            message: format!(
                "{}/{} 符合 Conventional Commits ({:.0}%)",
                compliant_count,
                commits.len(),
                compliance_rate
            ),
            suggestion: if !passed {
                Some("使用 `cz commit` 创建规范提交，或手动遵循 <type>: <description> 格式".to_string())
            } else {
                None
            },
        });
    }

    fn check_release_consistency(&mut self) {
        let changelog_path = self.repo_path.join("CHANGELOG.md");
        let changelog = match self.read_file(&changelog_path) {
            Some(c) => c,
            None => return,
        };

        // Try pyproject.toml first, then Cargo.toml
        let version = self
            .read_version_from_pyproject()
            .or_else(|| self.read_version_from_cargo());

        let pyproject_version = match version {
            Some(v) => v,
            None => return,
        };

        let changelog_has_version = Regex::new(&format!(
            r"## \[?{}]?",
            regex::escape(&pyproject_version)
        ))
        .unwrap()
        .is_match(&changelog);

        let has_version_commit = git_utils::run_git_cmd(&["log", "--oneline", "-20"], &self.repo_path)
            .map(|out| {
                let pattern = format!(
                    r"bump.*{}|v{}",
                    regex::escape(&pyproject_version),
                    regex::escape(&pyproject_version),
                );
                Regex::new(&pattern)
                    .ok()
                    .map(|re| re.is_match(&out.to_lowercase()))
                    .unwrap_or(true)
            })
            .unwrap_or(true);

        let passed = changelog_has_version && has_version_commit;
        self.add_result(AuditResult {
            name: "版本发布规范一致性".to_string(),
            passed,
            message: if passed {
                "CHANGELOG 和版本文件一致，且有版本提交".to_string()
            } else {
                format!("CHANGELOG 缺少 v{pyproject_version} 或缺少版本提交")
            },
            suggestion: if !passed {
                Some("发布前确保：1) 更新 CHANGELOG.md 2) 更新版本文件 3) 提交版本更新".to_string())
            } else {
                None
            },
        });
    }

    fn read_version_from_pyproject(&self) -> Option<String> {
        let path = self.repo_path.join("pyproject.toml");
        let content = self.read_file(&path)?;
        let re = Regex::new(r#"version\s*=\s*"([^"]+)""#).ok()?;
        re.captures(&content)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().to_string())
    }

    fn read_version_from_cargo(&self) -> Option<String> {
        let path = self.repo_path.join("Cargo.toml");
        let content = self.read_file(&path)?;
        let re = Regex::new(r#"version\s*=\s*"([^"]+)""#).ok()?;
        re.captures(&content)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().to_string())
    }
}

pub fn run(args: &AuditArgs) {
    let mut auditor = GitRepoAuditor::new(&args.repo_path);
    let report = auditor.audit();
    let passed = report.print_report(args.verbose);
    if !passed {
        std::process::exit(1);
    }
}
